Tan Jean Ren, Adriel iP

1st2GetThisName opened the pull request:
# Vecrosen
## A sloppily made memo app. 😛 
**It can store:**
- Todos ✅ 
- Deadlines 📆 
- Events 🎉 

*It looks something like this:*
1. [T][X] Breathe
2. [D][ ] Submit homework (by: 17 Jan 2026)
3. [E][ ] Halloween month (begin: 01 Oct 2025, end: 31 Oct 2026)

*And functions something like this:*
- [X] Breathe
- [ ] Submit homework (by: 17 Jan 2026)
- [ ] Halloween month (begin: 01 Oct 2025, end: 31 Oct 2026)

All made possible with our in-house parser `Parser.java`!
Here's a preview of it!
```java
        } else if (input.matches("event .+ /end .+ /begin .+")) {
            int beginStart = input.indexOf("/begin ");
            int endStart = input.indexOf("/end ");
            String desc = input.substring(6, endStart - 1);
            String end = input.substring(endStart + 5, beginStart - 1);
            String begin = input.substring(beginStart + 7);
            returnArgs.add(desc);
            returnArgs.add(begin);
            returnArgs.add(end);
            return ActionType.event;
```
> Made possible with [Duke](https://github.com/nus-cs2103-AY2425S1/ip) as a template.

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

1st2GetThisName committed:
Rename Duke.java to Vecrosen.java and Add greeting

The chatbot is named Duke and has Duke's greeting.

The chatbot needs to be distinguished from Duke.

Let's change the name to Vecrosen and replace the greetings.

1st2GetThisName committed:
Add indentation and echoing

Vecrosen does not do anything other than greet the user.
Vecrosen's text is difficult to differentiate from the user's.

Let's make the chatbot parrot the user and add 4 spaces before
anything Vecrosen's says.

Also, apparently the class name within the Vecrosen.java only got
refactored now? Visual Studio's being weird.

1st2GetThisName committed:
Add special input: "bye"

Vecrosen has no way of terminating naturally.

Let's allow the user to end the interaction by entering "bye".

1st2GetThisName committed:
Add special input: "list"

Vecrosen is still lacking in functionality.

Let's add a function for Vecrosen to output a list of items.

1st2GetThisName committed:
Add support for updating the list

The list that Vecrosen provides is static.
The speak() function doesn't allow text to be added before custom
inputs without using string concatenation.

If the user's input is large, string concatenation would have a large
time complexity, even if the string we wish to add is short.

Let's add anything the user says to the list and update speak() to
accept a prefix which is printed before the text.

1st2GetThisName committed:
Add state: "done"

The items in the list cannot be modified after being added.

Let's add a boolean to each list item to represent whether it has been
completed, and update the output of "list" to show if it has.

1st2GetThisName committed:
Add special input: "mark X"

The "done" variable has no functionality, it instead alternates
value on creation.

Let's instead allow users to manually mark tasks as done by
specifying the index of the desired task after the word "mark".

1st2GetThisName committed:
Change implementation of tasks to being their own objects

The list of tasks is stored as an arraylist of strings and an
arraylist of booleans.

As the functionality of tasks is expected to grow, it is better
to treat each task as an object moving forward.

Let's create a Task class and update the implementation to
utilize it.

1st2GetThisName committed:
Add special input: "unmark X" and fix numbering bug

Tasks, once marked as complete, could not be unmarked.
The numbering started at 46 instead of having a period
added after the number.

Let's add the command unmark and fix the aforementioned bug.

1st2GetThisName committed:
Add the ability to specify task types

All non-special input is treated as a task, hence only one
kind of task exists.

Let's allow the user to specify the type of task with the first word.

1st2GetThisName committed:
Add support for Deadlines and Events

Deadlines and Events are just tasks with details appended to
the back of their descriptions.

Let's make subclasses of Task so that the details can be stored
in variables.

1st2GetThisName committed:
Add minor error handling

When Vecrosen receives input in an invalid format, it does
not inform the user of how to properly use the command.

Let's add explanations as to how to use each command.

1st2GetThisName committed:
Add special input: "delete"

Users have no way of deleting tasks once they have been added to the list,
they could only mark them as complete.

Let's add a delete function to allow users to delete tasks.

1st2GetThisName committed:
Add testcase for automated text-UI testing

Testing has to be done manually.

Let's add a testcase so that the supplied runtest files can be used.

1st2GetThisName committed:
Fix bugs with the greeting and testcase

1st2GetThisName committed:
Fix bugs regarding Scanner

Scanner is initialized within the while(true) loop.

Hence, a new instance of Scanner is made on each loop.

Let's move Scanner out of the while loop.

(Also, fixed a typo)

1st2GetThisName committed:
Add save functionality

The list ceases to exist when the user exits the program.

Let's record the user's list in a file named vecrosen.txt.

1st2GetThisName committed:
Add load functionality

The data in vecrosen.txt is not accessed by the program, and is
not formatted to be read easily by a Scanner.

Let's modify the save format and add the ability to load the file's contents.

1st2GetThisName committed:
Merge branch 'branch-Level-7'

1st2GetThisName committed:
Add date recognition for deadlines

Deadlines are stored as strings rather than dates.

While this is more flexible in some cases, in other cases it looks lazy.

Let's add date recognition to the by field.

I'm not here to jostle with java's DateTimeFormatter object. I. Am. DONE.

1st2GetThisName committed:
Extend date detection to Events

Only the field "by" under deadlines detect dates.

Let's make date detection into its own class and implement it into Event.java.

1st2GetThisName committed:
Merge branch 'branch-Level-8'

1st2GetThisName committed:
Make UI its own class

Code is currently all run under a single program Vecrosen.java

Modularizing the code has benefits down the line.

Let's create a class called UI.

1st2GetThisName committed:
Make parsing its own class

Let's add a new class Parser to handle parsing.

1st2GetThisName committed:
Make savefile access its own class

Let's add a class called Storage to handle reading and writing to the savefile.

1st2GetThisName committed:
Make the tasklist its own class

Let's make a class called TaskList which manages functions updating and reading the list.

1st2GetThisName committed:
Merge branch 'branch-A-MoreOOP'

1st2GetThisName committed:
Move all files into a package named vecrosen

The files are in the source root folder.

Let's create a package to store them.

Dividing them into smaller packages can wait.

1st2GetThisName committed:
Merge branch 'add-gradle-support'

1st2GetThisName committed:
Add JUnit Tests

Let's add testcases to make use of Gradle.

1st2GetThisName committed:
Add shadowJar support

The program cannot be exported.

Let's update build.gradle to allow for .jar files to be created.
Let's also update Vecrosen to make the directory data if it doesn't yet exist.

1st2GetThisName committed:
Add Javadoc comments

Some functions have no Javadoc comments

Let's add Javadoc comments to the remaining functions

Bugfixes: Vecrosen.java now reads the user's input through Ui.java
Adjustment from 1-indexing to 0-indexing moved to TaskList.java

1st2GetThisName committed:
Make code compliant with coding standard

Code does not comply with coding standard

Let's make the necessary adjustments to make it so.

1st2GetThisName committed:
Add special input: find

Users cannot filter the list.

Let's add a find function to allow users to narrow their list down by a keyword.

1st2GetThisName committed:
Merge branch 'branch-A-Javadoc'

1st2GetThisName committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/vecrosen/Parser.java

1st2GetThisName committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/vecrosen/Parser.java
#	src/main/java/vecrosen/TaskList.java
#	src/main/java/vecrosen/Ui.java

sooyitao commented on src/main/java/vecrosen/CustomDateTimeParser.java:
```diff
@@ -0,0 +1,38 @@
+ */
+public class CustomDateTimeParser {
+    private static final String[] supportedDates = {
+            "uuuu-MM-dd", "uu-MM-dd",
+            "dd MMM, uuuu", "MMM dd, uuuu", "dd MMM uuuu", "MMM dd uuuu", "uuuu MMM dd", "uuuu dd MMM",
+            "dd MMMM, uuuu", "MMMM dd, uuuu", "dd MMMM uuuu", "MMMM dd uuuu", "uuuu MMMM dd", "uuuu dd MMMM",
+            "MM/dd/uuuu", "MM/dd/uu", "MM/dd",
+            "MMM dd", "dd MMM",
+            "MMMM dd", "dd MMMM"
```
Should the indentation be at 8 instead of 12?

1st2GetThisName replied:
I figured it should be at 12 since this is a wrapped line, even though the parent line ended in a {.

sooyitao commented on src/main/java/vecrosen/Event.java:
```diff
@@ -0,0 +1,72 @@
+    private LocalDate beginDate;
+    private LocalDate endDate;
+
+    /**
+     * Initializes a new Event object. Starts marked as incomplete.
+     * @param description The description of the event
+     * @param begin The start date/time of the event
+     * @param end The end date/time of the event
+     */
```
Perhaps there should be an empty line between the description and parameter section? I noticed the same issue in several other places too!

sooyitao commented on src/main/java/vecrosen/Storage.java:
```diff
@@ -0,0 +1,95 @@
+                int tth = s.nextInt();
+                TaskType tt = TaskType.values()[tth];
+                s.nextLine();
+                switch (tt) {
+                case event:
+                    String begin = s.nextLine();
+                    String end = s.nextLine();
+                    list.add(new Event(description, begin, end));
+                    break;
+                case deadline:
+                    String by = s.nextLine();
+                    list.add(new Deadline(description, by));
+                    break;
+                case todo:
+                    list.add(new Task(description));
+                    break;
+                default:
+                    System.err.println("TaskType not recognized when loading");
```
Great job on following coding standards for `switch case`!

sooyitao commented on src/main/java/vecrosen/Event.java:
```diff
@@ -0,0 +1,72 @@
+            beginDate = CustomDateTimeParser.parseDateTime(begin);
+            isDateBegin = true;
+        } catch (DateTimeParseException ignored) {
+            // no action
```
Good use of comment to explain code!

sooyitao reviewed (commented):
LGTM. Just a few minor coding standard related issues to fix. Overall the code was easy to read and follow. Good job!

realqijun commented on src/main/java/vecrosen/Ui.java:
```diff
@@ -0,0 +1,68 @@
+     * Obtains the next line that the user had entered
+     * @return The line the user entered
+     */
+    public String readline() {
```
Doesn't the l in readline require to be capitalized? I have seen readline in other languages not capitalized too, but coding standards says that we have to right?

realqijun commented on src/main/java/vecrosen/Deadline.java:
```diff
@@ -0,0 +1,49 @@
+    @Override
+    public String toString() {
+        String prev = super.toString();
+        return "[D" + prev.substring(2) + " (by: " + getBy() + ")";
```
I can't tell what prev means here, maybe you can change it to something like taskDesc or something else?

realqijun commented on src/main/java/vecrosen/Ui.java:
```diff
@@ -0,0 +1,68 @@
+     * and elaborates on how they were supposed to use said command.
+     * @param usage The proper usage of the command.
+     */
+    public void invalidFormat(String usage) {
```
I think this function name can be changed to something like alertInvalidFormat.

realqijun reviewed (commented):
Naming conventions are good to go!
Minor suggestions suggested, LGTM ☑️

1st2GetThisName committed:
Fix code formatting

There are suggestions on how to improve the code's formatting.

Let's implement them.

(and use Checkstyle while we're at it)

1st2GetThisName committed:
Add a basic dialog window

Vecrosen accepts inputs from console and prints to console.

Let's begin work on a dialog window.

1st2GetThisName committed:
Get dialog window working

Dialog window only shows a label

Let's make it work as if it were the console.

This has created a ton of style errors, but I am too fatigued and too overdue to fix it. So it's done.

1st2GetThisName committed:
Merge branch 'branch-Level-10'

1st2GetThisName committed:
Fix checkstyle issues

Checkstyle issues were created over the course of the previous updates.
Checkstyle flags array initialization as distinct from line extension
and hence expects array initialization to use 4 spaces more
instead of 8.

Let's fix the issues flagged by checkstyle except for indentation
during array initialization and update checkstyle to expect lines
in array initialization to be indented 8 spaces more than the
parent.

What do you mean we didn't need to write full git commit
messages, do you know how long it takes to write these?

1st2GetThisName committed:
Uncomment introduction

Introduction is commented out.

Let's uncomment it.

1st2GetThisName committed:
Add token assert

The project calls for an assert to be inserted into the code.

As I do not have the foresight to infer what meaningful asserts I could
include, let's add one assert function to satisfy the requirement.

1st2GetThisName committed:
Shorten long methods

Some methods have lines of code far exceeding 30 lines.

While we are unable to reduce it below 30, the number of lines can be
significantly reduced.

Thus, let's add layers of abstraction to reduce the number of lines in
these methods.

1st2GetThisName committed:
Add readme explaining the lack of streams

Streams is an optional task in the iP description.

Let's add a readme clarifying our lack of intent of doing it.

1st2GetThisName committed:
Merge pull request #2 from 1st2GetThisName/branch-A-Assertions

Add token assert

1st2GetThisName committed:
Merge branch 'master' into A-Streams

1st2GetThisName committed:
Merge branch 'master' into branch-A-CodeQuality

1st2GetThisName committed:
Merge pull request #3 from 1st2GetThisName/branch-A-CodeQuality

Shorten long methods

1st2GetThisName committed:
Merge branch 'master' into A-Streams

1st2GetThisName committed:
Merge pull request #4 from 1st2GetThisName/A-Streams

Add readme explaining the lack of streams

1st2GetThisName committed:
Implement framework for Client

Vecrosen only supports remembering tasks

Let's prepare to add a new type of object for
Vecrosen to remember: Clients.

1st2GetThisName committed:
Add load/save/show functionality to ClientList

There is no way to view the list of clients,
nor save or load it.

Let's make the program read/write to a separate savefile
to obtain the list of clients
Let's make the "list" command show the list of clients.

1st2GetThisName committed:
Implement special input: "client"

There is no way to edit the list of clients via the app.

Let's implement the command "client" to allow users to add
clients into the list.

1st2GetThisName committed:
Add readme

Readme is only a placeholder.

Let's update the readme to include the application's specifications.
